packages/async-rewriter3/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+            value: true,
+        });
+    }
+
+    function _isp(p) {
+        return p && p[_syntheticPromise];
+    }
+
+    let _functionState = 'sync', _synchronousReturnValue, _ex;
+
+    const _asynchronousReturnValue = (async () => {
+    try {
+    "#
+)
+}
+
+fn make_end_fn_insertion(offset: TextSize) -> Insertion {
+    Insertion::new(
+        offset,
+        r#"
+        } catch (err) {
+            if (_functionState === 'sync') {
+                // Forward synchronous exceptions.
+                _synchronousReturnValue = err;
+                _functionState = 'threw';
+            } else {
+                throw err;
+            }
+            } finally {
+            if (_functionState !== 'threw') {
+                _functionState = 'returned';
+            }
+        }
+
+        })();
+
+        if (_functionState === 'returned') {
+            return _synchronousReturnValue;
+        } else if (_functionState === 'threw') {
+            throw _synchronousReturnValue;
+        }
+
+        _functionState = 'async';
+        return _markSyntheticPromise(_asynchronousReturnValue);
+        "#
+    )
+}
+
+fn fn_start_insertion(body: &ExprOrBlock) -> InsertionList {
+    let mut ret = InsertionList::new();
+    let mut offset = body.syntax().text_range().start();
+    if !is_block(body) {
+        ret.push_back(Insertion::new(offset, "{"));
+    } else {
+        offset = offset.checked_add(1.into()).unwrap();
+    }
+    ret.push_back(make_start_fn_insertion(offset));
+    if !is_block(body) {
+        ret.push_back(Insertion::new(
+            offset,
+             "return ("
+        ));
+    }
+    ret
+}
+fn fn_end_insertion(body: &ExprOrBlock) -> InsertionList {
+    let mut ret = InsertionList::new();
+    let mut offset = body.syntax().text_range().end();
+    if is_block(body) {
+        offset = offset.checked_sub(1.into()).unwrap();
+    } else {
+        ret.push_back(Insertion::new(offset, ");"));
+    }
+    ret.push_back(make_end_fn_insertion(offset));
+    if !is_block(body) {
+        ret.push_back(Insertion::new(offset, "}"));
+    }
+    ret
+}
+
+fn add_all_variables_from_declaration(patterns: impl Iterator<Item = impl Borrow<Pattern>>) -> InsertionList {
+    let mut ret = InsertionList::new();
+    for pattern in patterns {
+        match pattern.borrow() {
+            Pattern::SinglePattern(p) => {
+                p.name().map(|name| ret.add_variable(name.to_string()));
+            },
+            Pattern::RestPattern(p) => {
+                let pat = p.pat();
+                if pat.is_some() {
+                    ret.append(&mut add_all_variables_from_declaration([&pat.unwrap()].into_iter()));
+                }
+            },
+            Pattern::AssignPattern(p) => {
+                let key = p.key();
+                if key.is_some() {
+                    ret.append(&mut add_all_variables_from_declaration([&key.unwrap()].into_iter()));
+                }
+            },
```

Comment:
You can also use match here to make it simpler to read and avoid that unwrap:

```rs
match p.key()  {
   Some(key) => ret.append(...key)
   _ => 
```
Or even an if with pattern extraction:

```rs
if let Some(key) = p.key {
  ret.append...
}
```